picasso/src/main/java/com/squareup/picasso3/NetworkRequestHandler.kt

```diff
@@ -0,0 +1,129 @@
+
+internal class NetworkRequestHandler(
+  private val callFactory: Call.Factory
+) : RequestHandler() {
+  override fun canHandleRequest(data: Request): Boolean {
+    val uri = data.uri ?: return false
+    val scheme = uri.scheme
+    return SCHEME_HTTP.equals(scheme, ignoreCase = true) ||
+        SCHEME_HTTPS.equals(scheme, ignoreCase = true)
+  }
+
+  override fun load(picasso: Picasso, request: Request, callback: Callback) {
+    val callRequest = createRequest(request)
+    callFactory
+      .newCall(callRequest)
+      .enqueue(object : okhttp3.Callback {
+        override fun onResponse(call: Call, response: Response) {
+          if (!response.isSuccessful) {
+            callback.onError(ResponseException(response.code, request.networkPolicy))
+            return
+          }
+
+          // Cache response is only null when the response comes fully from the network. Both
+          // completely cached and conditionally cached responses will have a non-null cache
+          // response.
+          val loadedFrom = if (response.cacheResponse == null) NETWORK else DISK
+
+          // Sometimes response content length is zero when requests are being replayed.
+          // Haven't found root cause to this but retrying the request seems safe to do so.
+          val body = response.body
+          if (loadedFrom == DISK && body!!.contentLength() == 0L) {
+            body.close()
+            callback.onError(
+              ContentLengthException("Received response with 0 content-length header.")
+            )
+            return
+          }
+          if (loadedFrom == NETWORK && body!!.contentLength() > 0) {
+            picasso.downloadFinished(body.contentLength())
+          }
+          try {
+            val bitmap = decodeStream(body!!.source(), request)
+            callback.onSuccess(Result.Bitmap(bitmap, loadedFrom))
+          } catch (e: IOException) {
+            body!!.close()
+            callback.onError(e)
+          }
+        }
+
+        override fun onFailure(call: Call, e: IOException) {
+          callback.onError(e)
+        }
+      })
+  }
+
+  override val retryCount: Int
+    get() = 2
+
+  override fun shouldRetry(airplaneMode: Boolean, info: NetworkInfo?): Boolean =
+    info == null || info.isConnected
+
+  override fun supportsReplay(): Boolean = true
+
+  private fun createRequest(request: Request): okhttp3.Request {
+    var cacheControl: CacheControl? = null
+    val networkPolicy = request.networkPolicy
+    if (networkPolicy != 0) {
+      cacheControl = if (isOfflineOnly(networkPolicy)) {
+        CacheControl.FORCE_CACHE
+      } else {
+        val builder = CacheControl.Builder()
+        if (!shouldReadFromDiskCache(networkPolicy)) {
+          builder.noCache()
+        }
+        if (!shouldWriteToDiskCache(networkPolicy)) {
+          builder.noStore()
+        }
+        builder.build()
+      }
+    }
+
+    val uri = checkNotNull(request.uri, "request.uri == null")
```

Comment:
nit: use Kotlin's native `checkNotNull` that doesn't allocate the error message String unless it's needed?